#165955694 Build out PATCH /todos/:todoId endpoint

frankly034 opened the pull request:
#### What does this PR do?
Have PATCH /todos/:todoId endpoint

#### Description of Task to be completed?
- add modify() to Todo model class 
- create PATCH /todos/:todoId route
- add verifyTokenMiddleware to PATCH /todos/:todoId and POST /todos routes
- add createTodo() to Validator class
- add createTodo middleware to POST /todos route

#### How should this be manually tested?
- Clone this branch
- Run ```npm run start-dev```
- Send a PATCH request to ```http://127.0.0.1:3000/api/v1/todos/:todoId```

#### Any background context you want to provide?
- You need node and npm installed
- Run ```npm install``` to install all dependencies 
- You need Postgres database
- Run ```cd src/db && sequelize db:migrate```
- Post data model ```
{
	"title": string,
	"description": string",
	"completed": boolean 
}```

#### What are the relevant pivotal tracker stories?
- #165955694 (https://www.pivotaltracker.com/story/show/165955694)

Commit:
feat(todo):update todo

- add model modify function
- add route PATCH /todos/:todoId

[Delivers #165955694]

Commit:
feat(todo):update todo

- add model modify function
- add route PATCH /todos/:todoId

[Delivers #165955694]

Commit:
feat(todo):update todo

- add model modify function
- add route PATCH /todos/:todoId

[Delivers #165955694]

Commit:
feat(todo):update todo

- add model modify function
- add route PATCH /todos/:todoId

[Delivers #165955694]

Commit:
> chore (verify Authentication)
> - Check Authorization
> [Delivers #166035752]

Commit:
> chore (verify Authentication)
> - Check Authorization
> [Delivers #166035752]

Commit:
feat(todo):validate create todo endpoint

- add createTodo validator function to validator.js
- add createTodo validation to POST /todos route

[delivers #165955694]

Commit:
Merge branch 'ft-edit-todo-endpoint-165955694' of https://github.com/frankly034/todo-app into ft-edit-todo-endpoint-165955694

Commit:
Merge branch 'ft-edit-todo-endpoint-165955694' of https://github.com/frankly034/todo-app into ft-edit-todo-endpoint-165955694

mosinmiloluwa01 commented:
hi Lewis,

Nice work so far, well done.

But i think you can check if a todo exists and then update it using todo.update instead of doing find by pk.

Also i think you are to verify that it is that user that owns the todo before u proceed to update the todo.

frankly034 commented:
Thanks @mosinmiloluwa01. 

I'll get right to it.

Commit:
feat(todo): create /todos/:id/completed routes

- create patch /todos/:id/completed routes
- protect routes
- validate todo id
- validate todo id against user id
- respond with a json obj

[Delivers #165957439]

Commit:
feat(todo): create /todos/:id/completed routes

- create patch /todos/:id/completed routes
- protect routes
- validate todo id
- validate todo id against user id
- respond with a json obj

[Delivers #165957439]

Commit:
Merge branch 'ft-update-todo-completed-#165957439' of https://github.com/frankly034/todo-app into ft-update-todo-completed-#165957439

Sojisoyoye committed:
feature(login): implement log in endpoint
- validate email input
- add convert email to lowercase
- add validations for password
[finishes #165952409]

Commit:
Merge branch 'ft-edit-todo-endpoint-165955694' of https://github.com/frankly034/todo-app into ft-edit-todo-endpoint-165955694

Commit:
Merge branch 'ft-edit-todo-endpoint-165955694' of https://github.com/frankly034/todo-app into ft-edit-todo-endpoint-165955694

Commit:
Merge branch 'ft-edit-todo-endpoint-165955694' of https://github.com/frankly034/todo-app into ft-edit-todo-endpoint-165955694

Commit:
Merge branch 'ft-edit-todo-endpoint-165955694' of https://github.com/frankly034/todo-app into ft-edit-todo-endpoint-165955694

EmmanuelSage reviewed (commented):
Nice one Lewis,
I can see the changes you made,
would you remove the completed rule from the validation as you are not using it again?

Asides the above question, it looks good to me

Sojisoyoye reviewed (commented):
@frankly034, good job...looks cool

Eazybee committed:
Merge pull request #12 from frankly034/ft-user-login-api-165952409

#165952409 Implement log in endpoint

Eazybee committed:
Merge branch 'develop' into ft-update-todo-completed-#165957439

Eazybee committed:
Merge pull request #11 from frankly034/ft-update-todo-completed-#165957439

#165957439 create /todos/:id/completed routes

Commit:
Merge branch 'ft-edit-todo-endpoint-165955694' of https://github.com/frankly034/todo-app into ft-edit-todo-endpoint-165955694

Commit:
Merge branch 'ft-edit-todo-endpoint-165955694' of https://github.com/frankly034/todo-app into ft-edit-todo-endpoint-165955694

Eazybee commented on src/server/controllers/todo.js:
```diff
@@ -46,5 +46,34 @@ class Todos {
+      }
+    })
+      .then(todo => {
+        if (!todo) {
```
i think we might need to test this point together because i have some  concern i can't really explain until testes

Eazybee commented on src/server/controllers/todo.js:
```diff
@@ -46,5 +46,34 @@ class Todos {
+      .then(updatedTodo => res.status(200).send({
+        status: 200,
+        message: 'Todo successfully updated',
+        data: updatedTodo
```
as well as this point

Eazybee commented on src/server/middlewares/validator.js:
```diff
@@ -32,5 +32,47 @@ class Validate {
+  }
+
+  static checkTodoIdParam (req, res, next) {
+    const todoId = req.params.todoId;
```
Optional Tips: you might want to use object destructuring here

akinyeleolat commented:
Once the merge conflict is sorted out, i think LGTM

Eazybee commented on src/server/middlewares/validator.js:
```diff
@@ -32,5 +32,47 @@ class Validate {
+    const rules = {
+      title: 'required|string', 
+      description: 'required|string', 
+      completed: 'required|max:5'
```
I don't think ```max:5``` should be here. 
completed is supposed to be boolean.

You might want to check validatorjs documentation for checking boolean.
Another opinion is all new todo should be false by default when created

Eazybee commented on src/server/middlewares/validator.js:
```diff
@@ -32,5 +32,47 @@ class Validate {
   }
+
+  static createTodo (req, res, next){
+    const { title, description, completed } = req.body;
```
I reviewed ```todoController.create()``` in  ```controllers/todo.js``` and found out that in your controller, you used ```const { userId } = req.params;```  but you didn't validate the param your ```createTodo()``` in ```validator.js```

frankly034 replied:
I've made the changes

Eazybee reviewed (changes requested):
Am really impressed by your pull request and I understand the logic behind it. However, there are some things I think you need to look into to give your code more quality

Commit:
Merge branch 'ft-edit-todo-endpoint-165955694' of https://github.com/frankly034/todo-app into ft-edit-todo-endpoint-165955694

Sojisoyoye commented:
@frankly034, Good job so far, especially by implementing all the feedbacks @Eazybee gave. Who is resolving the merge conflicts?

Eazybee committed:
Merge pull request #13 from frankly034/ch-check-authorization-166035752

#166035752 Check Authorization Token for frontend Implementation

Eazybee commented:
Please resolve conflict before merging

Commit:
Merge branch 'ft-edit-todo-endpoint-165955694' of https://github.com/frankly034/todo-app into ft-edit-todo-endpoint-165955694
